### PR TITLE
MNI-fsaverage Mapping

### LINKED ISSUE #622: Improved surface/volume map translation
https://github.com/ThomasYeoLab/CBIG/tree/master/stable_projects/registration/Wu2017_RegistrationFusion

### PR DESCRIPTION
**Pull request by aphroditepv**
Fix #622  

Files modified:
- apps/statmaps/tasks.py: added task generate_surface_image
- apps/statmaps/models.py: will now run generate_surface_image when new image is uploaded

Files added: mappings used in generate_surface_image
- apps/statmaps/static/anatomica/lh.avgMapping_allSub_RF_ANTs_MNI2fs.mat
- apps/statmaps/static/anatomica/lh.avgMapping_allSub_RF_ANTs_MNI2fs.mat

### CODE REVIEW

File: neurovault/apps/statmaps/tasks.py [hunk 1]
```diff
@@ -174,6 +174,65 @@ def generate_glassbrain_image(image_pk):
...
+        vox_coor = numpy.dot(numpy.linalg.inv(vox2ras[0:3, 0:3]), ras_centered)  # convert to voxel coordinates
+
+        img_surf = nib.gifti.GiftiImage()
+        for i in range(data_dim):
+            if data_vol.ndim > 3:
+                data_curr = data_vol[:, :, :, i]
+                data_surf = interpn([range(data_vol.shape[0]), range(data_vol.shape[1]), range(data_vol.shape[2])],
+                                    data_curr, vox_coor.T, 'linear')
+                data_surf_gifti = nib.gifti.GiftiDataArray(data_surf, 'NIFTI_INTENT_TIME_SERIES', 'NIFTI_TYPE_FLOAT32',
+                                                           'ASCII')
```
Comment (TimVanMourik) [inline neurovault/apps/statmaps/tasks.py]: I noticed this line and field 'ASCII'. I believe most data is currently using 'GZipBase64Binary' instead, so you might consider changing it. It should be automatically parsed by nibabel either way.

File: neurovault/apps/statmaps/tasks.py [hunk 2]
```diff
@@ -174,6 +174,65 @@ def generate_glassbrain_image(image_pk):
...
+    if data_vol.ndim > 3:
+        data_dim = data_vol.shape[3]  #number of time points
+    else:
+        data_dim = 1
+    this_path = os.path.abspath(os.path.dirname(__file__))
+
+    for hemi in ['lh', 'rh']:
+        if hemi == 'lh':
+            ras_coor = loadmat(os.path.abspath(os.path.join(this_path, "static", "anatomical",
+                                                            "lh.avgMapping_allSub_RF_ANTs_MNI2fs.mat")))['ras']
```
Comment (TimVanMourik) [inline neurovault/apps/statmaps/tasks.py]: Consider replacing the if statement by a string interpolation:
`"%s.avgMapping_allSub_RF_ANTs_MNI2fs.mat" % hemi`

**Comment by chrisfilo** [comment]
Thanks for your comments @TimVanMourik.

I worked on this on the plane back home - @aphroditepv I sent a pull request to your fork (see https://github.com/aphroditepv/NeuroVault/pull/1). I'm still tracking down a bug where surface files get created and then removed later on. Something is going wrong with concurrency.

File: neurovault/apps/statmaps/tasks.py (hunk 1 above)
Comment (chrisfilo) [reply neurovault/apps/statmaps/tasks.py]: I've tried with other encodings, but they produce outputs that seem to be incompatible with Connectome Workbench (see https://github.com/nipy/nibabel/issues/642). Besides 'ASCII' surprisingly produces the smallest outputs.

**Pull request closed without merging**